### DISCUSSION TITLE
Importing Into Set of String Data Type

### DISCUSSION THREAD

**Discussion by ramyabitorder** [SalesforceCommerceCloud/sfcc-ci#424] [Team Posts]
Hello All,

We have a Set of Strings Data Type in our Product object. We need to populate it programmatically. I am getting the values  and iterating as follows.

activityArrayToLoop.forEach(function(key ) {
	xmlString += '<value>'+key+'</value>'+'<br />';	
});

to generate a string as below. 
`<custom-attribute attribute-id="activity" xml:lang="x-default">
                <value>Snorkel + Scuba</value><br />
                <value>Travel</value><br />
                <value>Hike + Backpack</value><br />
                <value>Ski + Snowboard</value><br />
            </custom-attribute>`
  

When i import the same into the product object , all the values are populating in one field rather than multiple fields in the Set of Strings Data Type. I have tried adding ‘\n’ ‘<br><\br>’ and <br /> to insert new lines.  I have even tried putting the above into CDATA. But none of these have helped inserting them separately into product. 

When I import the xml from the instance i get below , so I believe Im building it correctly, however I am missing something to do with new line . Please suggest I am doing anything wrong here. 
`<custom-attribute attribute-id="activity" xml:lang="x-default">
                <value>Snorkel + Scuba</value>
                <value>Travel</value>
                <value>Hike + Backpack</value>
                <value>Ski + Snowboard</value>
            </custom-attribute>
`

**Comment by pravngaur** [comment]
Hi,

You need to insert an array rather than a string, just as you are getting
it from API

**Comment by ramyabitorder** [comment]
Hello Praveen Gaur,

Sorry for the delay.
Im using a FileWriter to write a xml file downloaded from a server. this is imported into the system programmatically , So i need to use a string format like below to import the xml into a 'set of strings' datatype field.

Im sure, Im missing something minor here, because when I manually import the file , the value within <values>  in the below string get populated correctly in my Custom Activity field.

<custom-attribute attribute-id="activity" xml:lang="x-default"> <value>Snorkel + Scuba</value> <value>Travel</value> <value>Hike + Backpack</value> <value>Ski + Snowboard</value> </custom-attribute>

**Comment by zlatinz** [comment]
Have you tried to do a manual edit and export of XML to see how exactly the XML looks like after the export?

**Comment by ramyabitorder** [comment]
Hello Zlatin,

It looks exactly like how I have built it when I export it from the system. Below is the one i just got from the exported xml again.

> 

`<custom-attribute attribute-id="activity" xml:lang="x-default">
                <value>Snorkel + Scuba</value>
                <value>Travel</value>
                <value>Hike + Backpack</value>
                <value>Ski + Snowboard</value>
            </custom-attribute>`

>

**Comment by zlatinz** [comment]
Hi @ramyabitorder,
Open the exported file and the one you generated in Notepad++ and from the menu select  View > Show Symbol > Show All Characters. This will show you the control characters, including CR and LF and will help you to visually distinguish between the two files.
Alternatively, in VSCode you can see if the file uses CRLF or LF style of line ending in the status bar when the file is open.
Also have you tried importing the exported file after you have cleared the attribute to see if it works properly?

**Comment by pravngaur** [comment]
Hi @ramyabitorder & Zlatin,

If I understood correctly then Ramya you are trying to insert the values in a set of string attribute on product object. And you are trying to do it programmatically but ending up in setting all the values as a single String literal in the first value in the attribute, as opposed to having it as different string values.

For setting the set of strings, arrays need to be passed as set-of-string attribute value, pseudo code:

require('dw/system/Transaction').wrap(function(){
            var config = require('dw/object/CustomObjectMgr').createCustomObject("TestCustomObject", "Games");
            config.custom.setOfStringTesting = ["Snorkel + Scuba", "Travel", "Hike + Backpack", "Ski + Snowboard"];
            
        });

This code works for me, have a look at the attached image - setting up the right values using the above code. I am somehow not able to attach the files here.
<img width="1106" alt="Screenshot 2019-07-16 at 10 57 57 AM" src="https://user-images.githubusercontent.com/42477285/61268283-9d83ea80-a7b8-11e9-914d-925a4a30f441.png">

Irrespective of the source data, if you trying to update set of string then arrays need to be used.

Both, let me know if I misunderstood it or Ramya if you face any issue in going through the code.

**Comment by zlatinz** [comment]
Hi @pravngaur,
The way I understood the question is that he tries to generate (programmatically) an XML file, which after it is imported updates a set-of-string attribute.

**Comment by ramyabitorder** [comment]
Thats correct. 
I am writing a all this data into XML file (like below), this xml fiel si eventually imported to set-of-strings field in the product object 
 

`<custom-attribute attribute-id="activity" xml:lang="x-default"> <value>Snorkel + Scuba</value> <value>Travel</value> <value>Hike + Backpack</value> <value>Ski + Snowboard</value> </custom-attribute>`

**Comment by zlatinz** [comment]
Hi @ramyabitorder 
please can you provide the code snippet you use to generate this XML fragment?
This may help identifying the issue quicker.

Thanks,
Zlatin

**Comment by ramyabitorder** [comment]
Hello @zlatinz 

Here is my code snippet. The following code is within a while loop. Please note this is not complete code , but just to give a brief of how its built. I have attribute map from which Im setting the values for the fields. Here Activity field is of type  set of strings. 
(sorry about the formatting, IM unable to get it rt)



`if (file.createNewFile()) {
            	var exportFileWriter : FileWriter = new FileWriter(file, "UTF-8");
            	exportFileWriter.write('<catalog xmlns="http://www.demandware.com/xml/impex/catalog/2006-10-31" catalog-id="'+StringUtils.stringToXml(catalogID)+'">\n');
				exportFileWriter.write('<header>\n');
				exportFileWriter.write('</header>\n');
				var productID : String = eachProject.getCustom()['ArtifactID'];
				var displayName : String = attributesMap.get("name");
				var productBrand : String = attributesMap.get("brand");
				var activity : String = attributesMap.get("activity”);//the data type of this field is Set-Of_strings. Contains <value>Snorkel + Scuba</value> <value>Travel</value> <value>Hike + Backpack</value> <value>Ski + Snowboard</value> 





if(productID != undefined && productID != null && productID != '')
				exportFileWriter.write('<product product-id="'+StringUtils.stringToXml(productID)+'">\n');
				exportFileWriter.write('<ean/>');
				exportFileWriter.write('<upc/>');
				exportFileWriter.write('<unit/>');
        		exportFileWriter.write('<min-order-quantity>'+StringUtils.stringToXml('1')+'</min-order-quantity>');
				exportFileWriter.write('<step-quantity>'+StringUtils.stringToXml('1')+'</step-quantity>');
				if(displayName != undefined && displayName != null && displayName != '')
				exportFileWriter.write('<display-name xml:lang="'+StringUtils.stringToXml(targetLocale)+'">'+StringUtils.stringToXml(displayName)+'</display-name>');
				if(shortDescription != undefined && shortDescription != null && shortDescription != '')
				exportFileWriter.write('<short-description xml:lang="'+StringUtils.stringToXml(targetLocale)+'">'+StringUtils.stringToXml(shortDescription)+'</short-description>');
				exportFileWriter.write('<online-flag>'+StringUtils.stringToXml('true')+'</online-flag>');
				exportFileWriter.write('<available-flag>'+StringUtils.stringToXml('true')+'</available-flag>');
				exportFileWriter.write('<searchable-flag>'+StringUtils.stringToXml('true')+'</searchable-flag>');
				exportFileWriter.write('<tax-class-id>'+StringUtils.stringToXml('standard')+'</tax-class-id>');

				if(productBrand != undefined && productBrand != null && productBrand != '')
				exportFileWriter.write('<brand>'+StringUtils.stringToXml(productBrand)+'</brand>');
if(activity != undefined && activity != null && activity != '')
				exportFileWriter.write('<custom-attribute attribute-id="activity" xml:lang="'+StringUtils.stringToXml(targetLocale)+'">'+StringUtils.stringToXml(activity)+'</custom-attribute>');


	exportFileWriter.write('</custom-attributes>');
				exportFileWriter.write('<pinterest-enabled-flag>'+StringUtils.stringToXml('false')+'</pinterest-enabled-flag>');
				exportFileWriter.write('<facebook-enabled-flag>'+StringUtils.stringToXml('true')+'</facebook-enabled-flag>');
				exportFileWriter.write('</product>');
				exportFileWriter.write('</catalog>\n');
				exportFileWriter.close();
`
}

**Comment by zlatinz** [comment]
Several recommendations:

1. Try using writeLine method instead of write (also do not use the \n character on your own). See https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/DWAPI/scriptapi/html/api/class_dw_io_FileWriter.html#dw_io_FileWriter_writeLine_String_DetailAnchor
2. Make sure to use .flush() method before .close()
3. When making this a production code, make sure to enclose FileWriter operations in try catch finally block. In finally block make sure to flush and close the FileWriter.

**Comment by pravngaur** [comment]
My bad!!

Just to add to what Zlatin said, please check if you can use XMLStreamWriter, that will wrap around the XML specific functions, which will take away a lot of low level stuff.

As @zlatinz suggested, flush and try-catch holds true to both the writers.

Let us know how it goes.
